content/en/blog/_posts/2025-01-14-devices-failure-handling/index.md

```diff
@@ -0,0 +1,409 @@
+accelerators, there are new failure modes. Device drivers installed on the node:
+
+* Must match the hardware  
+* Be compatible with an app  
+* Must work with other drivers (like [nccl](https://developer.nvidia.com/nccl),
+  etc.)
+
+Best practices handling driver versions:
+
+* Monitor driver installer health  
+* Plan upgrades of infrastructure and Pods to match the version  
+* Have canary deployments whenever possible
+
+Following the best practices in this section and using device plugins and device
+driver installers from trusted and reliable sources generally eliminate this
+class of failures. Kubernetes is tracking work to make this space even better.
+
+### Failure modes: device failed
+
+There is very little handling of device failure in Kubernetes today. Device
+plugins report the device failure only by changing the count of allocatable
+devices. And Kubernetes relies on standard mechanisms like liveness probes or
+container failures to allow Pods to communicate the failure condition to the
+kubelet. However, Kubernetes does not correlate device failures with container
+crashes and does not offer any mitigation beyond restarting the container while
+being attached to the same device.
+
+This is why many plugins and DIY solutions exist to handle device failures based
+on various signals.
+
+#### Health controller
+
+In many cases a failed device will result in unrecoverable and very expensive
+nodes doing nothing. A simple DIY solution is a _node health controller_. The
+controller could compare the device allocatable count with the capacity and if
+the capacity is greater, it starts a timer. Once the timer reaches a threshold,
+the health controller kills and recreates a node.
+
+There are problems with the _health controller_ approach:
+
+* Root cause of the device failure is typically not known  
+* The controller is not workload aware  
+* Failed device might not be in use and you want to keep other devices running  
+* The detection may be too slow as it is very generic  
+* The node may be part of a bigger set of nodes and simply cannot be deleted in
+  isolation without other nodes
+
+There are variations of the health controller solving some of the problems
+above. The overall theme here though is that to best handle failed devices, you
+need customized handling for the specific workload. Kubernetes doesn’t yet offer
+enough abstraction to express how critical the device is for a node, for the
+cluster, and for the Pod it is assigned to.
+
+#### Pod failure policy
+
+Another DIY approach for device failure handling is a per-pod reaction on a
+failed device. This approach is applicable for *training* workloads that are
+implemented as Jobs.
+
+Pod can define special error codes for device failures. For example, whenever
+unexpected device behavior is encountered, Pod exits with a special exit code.
+Then the Pod failure policy can handle the device failure in a special way. Read
```

Comment:
Probably `in a special way` could be elaborated a bit for a better understanding (I know the doc is linked below though). What should we do with those special error codes coming from device failures?

Reply:
I don't want to repeat documentation in the blog post. Or I misundertood the suggestion?